python_modules/dagster/dagster/_core/execution/submit_asset_runs.py

```diff
@@ -129,50 +130,53 @@ def _create_asset_run(
-                run_config={},
-                step_keys_to_execute=None,
-                tags=run_request.tags,
-                root_run_id=None,
-                parent_run_id=None,
-                status=DagsterRunStatus.NOT_STARTED,
-                external_job_origin=external_job.get_external_origin(),
-                job_code_origin=external_job.get_python_origin(),
-                asset_selection=frozenset(run_request.asset_selection),
-                asset_check_selection=None,
-                asset_job_partitions_def=partitions_def,
+        try:
+            # create a new request context for each run in case the code location server
+            # is swapped out in the middle of the submission process
+            workspace = workspace_process_context.create_request_context()
+            execution_data = _get_job_execution_data_from_run_request(
+                asset_graph,
+                run_request,
+                instance,
+                workspace=workspace,
+                run_request_execution_data_cache=run_request_execution_data_cache,
             )
+            check_for_debug_crash(debug_crash_flags, "EXECUTION_PLAN_CREATED")
+            check_for_debug_crash(debug_crash_flags, f"EXECUTION_PLAN_CREATED_{run_request_index}")
+
+            # retry until the execution plan targets the asset selection
+            if _execution_plan_targets_asset_selection(
+                execution_data.external_execution_plan.execution_plan_snapshot,
+                check.not_none(run_request.asset_selection),
+            ):
+                external_job = execution_data.external_job
+                external_execution_plan = execution_data.external_execution_plan
+                partitions_def = execution_data.partitions_def
+
+                run = instance.create_run(
+                    job_snapshot=external_job.job_snapshot,
+                    execution_plan_snapshot=external_execution_plan.execution_plan_snapshot,
+                    parent_job_snapshot=external_job.parent_job_snapshot,
+                    job_name=external_job.name,
+                    run_id=run_id,
+                    resolved_op_selection=None,
+                    op_selection=None,
+                    run_config={},
+                    step_keys_to_execute=None,
+                    tags=run_request.tags,
+                    root_run_id=None,
+                    parent_run_id=None,
+                    status=DagsterRunStatus.NOT_STARTED,
+                    external_job_origin=external_job.get_external_origin(),
+                    job_code_origin=external_job.get_python_origin(),
+                    asset_selection=frozenset(run_request.asset_selection),
+                    asset_check_selection=None,
+                    asset_job_partitions_def=partitions_def,
+                )
 
-            return run
+                return run
+        except DagsterInvalidSubsetError:
+            pass
```

Comment:
@clairelin135 This change addresses a very subtle issue with functionality addressed in #19091. That PR addresses the situation where, in the middle of a backfill, the targeted job is changed to no longer include the targeted asset. It solves this problem by examining the execution plan returned by `get_external_job` to see if it actually targets the requested asset-- if it does not, then the asset graph refreshes.

This approach was failing against the changes introduced in the present PR because trying to subset a job with an asset that is not targeted by that job now throws an `InvalidSubsetError` instead of returning a "null" execution plan. This was also true before, _but_ would never happen for base asset jobs because all other assets were included as source assets-- and the only reason we checked source assets is because observable sources could be in there. In this PR, we now only check the assets "targeted" by a job when subsetting (observables or materializables) and ignore assets used for loading only. So we now get a `DagsterInvalidSubsetError` in the situation of the original PR. I also think the execution plan approach also would've failed if targeting a non-base-asset job even prior to this PR, it might just have been a case we weren't testing for.

In any case, the change here is just to a `DagsterInvalidSubsetError` thrown by the code location and treat it in exactly the same way as when the execution plan doesn't match according to the logic you added.

Reply:
This looks reasonable to me!

Reply:
thanks for explaining